### ISSUE TITLE
weird behaviour of $|

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#2890]
Migrated from [rt.perl.org#4760](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=4760) (status was 'resolved')

Searchable as RT4760$

**Comment by p5pRT** [comment]
### From thospel@mail.dma.be

#### Created by thospel@mail.dma.be



Compare the output of these two programs&#8203;:
perl \-wle '          $| = ~$|; print $|'
perl \-wle 'my $a=$|; $| = ~$|; print $|'

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.6.0:

Configured by ton at Wed Aug  9 23:24:07 CEST 2000.

Summary of my perl5 (revision 5.0 version 6 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.4.0-test5, archname=i686-linux
    uname='linux quasar 2.4.0-test5 #2 sat jul 29 04:45:14 cest 2000 i686 unknown '
    config_args=''
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=undef 
    use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
  Compiler:
    cc='cc', optimize='-O2 -fomit-frame-pointer', gccversion=egcs-2.91.66 19990314 (egcs-1.1.2 release), gccosandvers=
    cppflags='-fno-strict-aliasing -I/usr/local/include'
    ccflags ='-fno-strict-aliasing -I/usr/local/include'
    stdchar='char', d_stdstdio=define, usevfork=false
    intsize=4, longsize=4, ptrsize=4, doublesize=8
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=4
    alignbytes=4, usemymalloc=y, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib -L/opt/gnu/lib'
    libpth=/usr/local/lib /opt/gnu/lib /lib /usr/lib
    libs=-lnsl -lndbm -ldl -lm -lc -lposix -lcrypt -lutil
    libc=/lib/libc-2.1.1.so, so=so, useshrplib=true, libperl=libperl.so
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic -Wl,-rpath,/usr/lib/perl5/5.6.0/i686-linux/CORE'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib -L/opt/gnu/lib'

Locally applied patches:
    


@INC for perl v5.6.0:
    /usr/lib/perl5/5.6.0/i686-linux
    /usr/lib/perl5/5.6.0
    /usr/lib/perl5/site_perl/5.6.0/i686-linux
    /usr/lib/perl5/site_perl/5.6.0
    /usr/lib/perl5/site_perl
    .


Environment for perl v5.6.0:
    HOME=/home/ton
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/ton/bin.Linux:/home/ton/bin:/home/ton/bin.SampleSetup:/usr/local/bin:/usr/local/sbin:/usr/local/jre/bin:/usr/local/jdk116_v5/bin:/usr/games/bin:/usr/X11R6/bin:/usr/share/bin:/usr/bin:/usr/sbin:/bin:/sbin:.
    PERL_BADLANG (unset)
    SHELL=/bin/bash


```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout

On Wed Nov 22 12&#8203;:06&#8203;:30 2000\, thospel@&#8203;mail\.dma\.be wrote&#8203;:
> This is a bug report for perl from thospel@&#8203;mail\.dma\.be\,
> generated with the help of perlbug 1\.31 running under perl v5\.6\.0\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
> Compare the output of these two programs&#8203;:
> perl \-wle '          $| = ~$|; print $|'
> perl \-wle 'my $a=$|; $| = ~$|; print $|'

Fixed by d8ce0c9a459\, according to the commit message and
\<http&#8203;://www\.nntp\.perl\.org/group/perl\.perl5\.porters/2000/11/msg26538\.html>\.

This is the cause of \#115206\.  I also suspect that this bug had a
different cause originally\, involving missing FETCH calls\, that has
subsequently been fixed\, because reverting it fixes \#115206 without
reintroducing this bug\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

On Tue Oct 09 20&#8203;:43&#8203;:09 2012\, sprout wrote&#8203;:
> On Wed Nov 22 12&#8203;:06&#8203;:30 2000\, thospel@&#8203;mail\.dma\.be wrote&#8203;:
> > This is a bug report for perl from thospel@&#8203;mail\.dma\.be\,
> > generated with the help of perlbug 1\.31 running under perl v5\.6\.0\.
> > 
> > 
> > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> > \[Please enter your report here\]
> > 
> > Compare the output of these two programs&#8203;:
> > perl \-wle '          $| = ~$|; print $|'
> > perl \-wle 'my $a=$|; $| = ~$|; print $|'
> 
> Fixed by d8ce0c9a459\, according to the commit message and
> \<http&#8203;://www\.nntp\.perl\.org/group/perl\.perl5\.porters/2000/11/msg26538\.html>\.
> 
> This is the cause of \#115206\.  I also suspect that this bug had a
> different cause originally\, involving missing FETCH calls\, that has
> subsequently been fixed\, because reverting it fixes \#115206 without
> reintroducing this bug\.

And I added a test for it in 8f25f6c03\.

\-\- 

Father Chrysostomos
